java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -242,116 +301,164 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getOutputInfo
+  for (int i = 0; i < numInputs; i++) {
+    javaInputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, inputNamesArr, i);
+    inputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaInputStrings[i], NULL);
+    inputValues[i] = (OrtValue*)inputTensors[i];
+  }
+
+  // Release the java array copy of pointers to the tensors.
+  (*jniEnv)->ReleaseLongArrayElements(jniEnv, tensorArr, inputTensors, JNI_ABORT);
+
+  // Extract the names of the output values.
+  for (int i = 0; i < numOutputs; i++) {
+    javaOutputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, outputNamesArr, i);
+    outputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaOutputStrings[i], NULL);
+    outputValues[i] = NULL;
+  }
 
-    // Release the Java input strings
-    for (int i = 0; i < numInputs; i++) {
-        (*jniEnv)->ReleaseStringUTFChars(jniEnv,javaInputStrings[i],inputNames[i]);
+  // Actually score the inputs.
+  // ORT_API_STATUS(OrtRun, _Inout_ OrtSession* sess, _In_ OrtRunOptions* run_options,
+  // _In_ const char* const* input_names, _In_ const OrtValue* const* input, size_t input_len,
+  // _In_ const char* const* output_names, size_t output_names_len, _Out_ OrtValue** output);
+  code = checkOrtStatus(jniEnv, api, api->Run(session, runOptions, (const char* const*)inputNames,
+                                              (const OrtValue* const*)inputValues, numInputs,
+                                              (const char* const*)outputNames, numOutputs, outputValues));
+  if (code != ORT_OK) {
+    goto cleanup_output_values;
+  }
+
+  // Construct the output array of ONNXValues
+  char* onnxValueClassName = "ai/onnxruntime/OnnxValue";
+  jclass onnxValueClass = (*jniEnv)->FindClass(jniEnv, onnxValueClassName);
+  outputArray = (*jniEnv)->NewObjectArray(jniEnv, safecast_int64_to_jsize(numOutputs), onnxValueClass, NULL);
+
+  // Convert the output tensors into ONNXValues
+  for (int i = 0; i < numOutputs; i++) {
+    if (outputValues[i] != NULL) {
+      jobject onnxValue = convertOrtValueToONNXValue(jniEnv, api, allocator, outputValues[i]);
+      if (onnxValue == NULL) {
+        break;  // go to cleanup, exception thrown
+      }
+      (*jniEnv)->SetObjectArrayElement(jniEnv, outputArray, i, onnxValue);
     }
+  }
 
-    // Release the buffers
-    checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, (void*)inputNames));
-    checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, (void*)inputValues));
-    checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, (void*)outputNames));
-    checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, javaInputStrings));
-    checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, javaOutputStrings));
+  // Note these gotos are in a specific order so they mirror the allocation pattern above.
+  // They must be changed if the allocation code is rearranged.
+cleanup_output_values:
+  checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, outputValues));
```

Comment:
>heckOrtStatus(jniEnv, api, api->AllocatorFree(allocator, outputValues)); [](http://example.com/codeflow?start=3&length=72)

Each individual OrtValue in the outputs if present, needs to be released via an API.